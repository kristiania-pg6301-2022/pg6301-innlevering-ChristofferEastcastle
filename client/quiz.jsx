import {
  BrowserRouter,
  Link,
  Route,
  Routes,
  useNavigate,
} from "react-router-dom";
import { useEffect, useState } from "react";
import * as React from "react";
import axios from "axios";

function FrontPage({ answered, correctAnswers }) {
  return (
    <div>
      <h1>QuizApplication</h1>
      <h3>Correct answers: {correctAnswers}</h3>
      <h3>Total answered: {answered}</h3>
      <ul>
        <li>
          <Link to="/question">Answer question</Link>
        </li>
      </ul>
    </div>
  );
}

export function QuestionPage({ answered, correctAnswers }) {
  const [question, setQuestion] = useState();

  const navigate = useNavigate();

  async function handleAnswer(answer) {
    answered((a) => a + 1);
    navigate("/answer/correct");

    const correct = await axios.post("/api/question", {
      id: question.id,
      answer: answer,
    });

    if (correct.data === true) {
      correctAnswers((c) => c + 1);
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }

  async function fetchQuestion() {
    return axios.get("/api/question");
  }

  useEffect(async () => {
    const text = await fetchQuestion();
    setQuestion(text.data);
  }, []);

  return (
    <div>
      <h1>Question</h1>
      <p>**Click on the one you think is right**</p>
      <Question
        question={question}
        handleAnswer={handleAnswer}
        answered={answered}
        correctAnswer={correctAnswers}
      />
    </div>
  );
}

export function Question({ question, handleAnswer }) {
  if (!question) return <h4>Loading...</h4>;

  return (
    <div>
      <h4>{question.question}</h4>
      {Object.keys(question.answers)
        .filter((k) => question.answers[k] != null)
        .map((k) => (
          <button style={{ margin: 5 }} onClick={() => handleAnswer(k)} key={k}>
            {question.answers[k]}
          </button>
        ))}
    </div>
  );
}

export function QuizApplication() {
  const [answered, setAnswered] = useState(0);
  const [correctAnswers, setCorrectAnswers] = useState(0);
  return (
    <BrowserRouter>
      <Routes>
        <Route
          path="/"
          element={
            <FrontPage answered={answered} correctAnswers={correctAnswers} />
          }
        />
        <Route
          path="/question"
          element={
            <QuestionPage
              answered={setAnswered}
              correctAnswers={setCorrectAnswers}
            />
          }
        />
        <Route
          path="/answer/correct"
          element={
            <div>
              <h1>Correct!</h1>
              <GoHomeOrNew />
            </div>
          }
        />
        } />
        <Route
          path="/answer/wrong"
          element={
            <div>
              <h1>Wrong answer...</h1>
              <GoHomeOrNew />
            </div>
          }
        />
        } />
      </Routes>
    </BrowserRouter>
  );
}

function GoHomeOrNew() {
  return (
    <div>
      <ul>
        <li>
          <Link to={"/question"}>New Question</Link>
        </li>

        <li>
          <Link to={"/"}>Go home</Link>
        </li>
      </ul>
    </div>
  );
}
