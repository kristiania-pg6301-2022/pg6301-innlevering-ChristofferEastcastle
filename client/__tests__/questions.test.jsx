import * as React from "react";
import ReactDOM from "react-dom";

import { Question, QuizApplication } from "../quiz.jsx";
import { act } from "react-dom/test-utils";

describe("quiz application", () => {


  it("shows front page", () => {
    const element = document.createElement("div");
    ReactDOM.render(<QuizApplication />, element);

    expect(element.querySelector("h1").innerHTML)
      .toEqual("QuizApplication");
    expect(element.innerHTML).toMatchSnapshot();
  });

  const mockQuestion = {
    id: 913,
    question: "Which of the following are core Kubernetes objects?",
    description: null,
    answers: {
      answer_a: "Pods, Services, Volumes",
      answer_b: "Pods, Docker, Volumes",
      answer_c: "Pods, Services, Droplets",
      answer_d: "None of the mentioned",
      answer_e: null,
      answer_f: null
    }
  };


  it("shows mock question", () => {

    const element = document.createElement("div");

    act(() => {
      ReactDOM.render(<Question question={mockQuestion} correctAnswer={0} answered={0} />, element);
    });
    expect(element.querySelector("h4").innerHTML)
      .toEqual(mockQuestion.question);
    expect(element.innerHTML).toMatchSnapshot();
  });
});
