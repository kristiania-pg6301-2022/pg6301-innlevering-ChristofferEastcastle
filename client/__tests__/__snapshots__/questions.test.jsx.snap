// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz application shows front page 1`] = `"<div><h1>QuizApplication</h1><h3>Correct answers: 0</h3><h3>Total answered: 0</h3><ul><li><a href=\\"/question\\">Answer question</a></li></ul></div>"`;

exports[`quiz application shows mock question 1`] = `"<div><h4>Which of the following are core Kubernetes objects?</h4><button style=\\"margin: 5px;\\">Pods, Services, Volumes</button><button style=\\"margin: 5px;\\">Pods, Docker, Volumes</button><button style=\\"margin: 5px;\\">Pods, Services, Droplets</button><button style=\\"margin: 5px;\\">None of the mentioned</button></div>"`;
